File: csharp/zugferd_facturx_xrechnung/attach_invoice/attach_invoice.cs
//  attach_invoice -- C# port of examples\Vb6\zugferd_facturx_xrechnung\attach_invoice\attach_invoice.bas
//  Imports an existing PDF/A-3 invoice, attaches the factur-x.xml e-invoice,
//  associates it with the catalog and sets the FacturX Comfort PDF version.
using System;
using System.IO;
using LumasPdfSdk;

class AttachInvoice
{
    // We try to continue if an error occurs
    static int PDFError(IntPtr Data, int ErrCode, string ErrMessage, int ErrType)
    {
        Console.WriteLine(ErrMessage);
        return 0;
    }

    static TErrorProc _errCb = PDFError;

    static void Main()
    {
        string tf = System.IO.Path.Combine(System.AppContext.BaseDirectory, "..", "..", "..", "..", "test_files");

        IntPtr pdf = LumasPdf.pdfNewPDF();
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, _errCb);
        LumasPdf.pdfCreateNewPDFW(pdf, "");        // The output file is opened later

        // We assume that the pdf invoice is already a valid PDF/A 3 file in this example.
        LumasPdf.pdfSetImportFlags(pdf, LumasPdfConsts.ifImportAsPage | LumasPdfConsts.ifImportAll);
        LumasPdf.pdfOpenImportFileW(pdf, tf + "test_invoice.pdf", (int)LumasPdfConsts.ptOpen, "");

        LumasPdf.pdfImportPDFFile(pdf, 1, 1.0, 1.0);

        int ef = LumasPdf.pdfAttachFileW(pdf, tf + "factur-x.xml", "EN 16931 compliant invoice", false);
        LumasPdf.pdfAssociateEmbFile(pdf, TAFDestObject.adCatalog, -1, TAFRelationship.arAlternative, (uint)ef);

        // ZUGFeRD 2.1+ and FacturX are identically defined in PDF and share the same version constants.
        LumasPdf.pdfSetPDFVersion(pdf, LumasPdfConsts.pvFacturX_Comfort);

        // No fatal error occurred?
        if (LumasPdf.pdfHaveOpenDoc(pdf))
        {
            // We write the file into the application directory.
            string outFile = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "out.pdf");
            if (!LumasPdf.pdfOpenOutputFileW(pdf, outFile))
            {
                LumasPdf.pdfDeletePDF(pdf);
                return;
            }
            if (LumasPdf.pdfCloseFile(pdf))
                Console.WriteLine("PDF file \"" + outFile + "\" successfully created!");
        }
        LumasPdf.pdfDeletePDF(pdf);
    }
}

File: cpp/zugferd_facturx_xrechnung/attach_invoice/attach_invoice.cpp
// attach_invoice -- C++ mirror of
// examples\Vb6\zugferd_facturx_xrechnung\attach_invoice\attach_invoice.bas
// Imports an existing PDF/A-3 invoice, attaches factur-x.xml, associates it with
// the catalog and sets the FacturX Comfort PDF version.
#include "apputil.h"

static SI32 PDF_CALL PDFError(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType) {
    if (ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

int main() {
    ChdirToExe();
    PPDF pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, 0, PDFError);
    pdfCreateNewPDFA(pdf, "");

    // We assume the invoice is already a valid PDF/A-3 file.
    pdfSetImportFlags(pdf, ifImportAsPage | ifImportAll);
    pdfOpenImportFileA(pdf, "../../../test_files/test_invoice.pdf", ptOpen, "");
    pdfImportPDFFile(pdf, 1, 1, 1);

    int ef = pdfAttachFileA(pdf, "../../../test_files/factur-x.xml", "EN 16931 compliant invoice", 0);
    pdfAssociateEmbFile(pdf, adCatalog, -1, arAlternative, ef);

    // ZUGFeRD 2.1+ and FacturX share the same PDF version constants.
    pdfSetPDFVersion(pdf, pvFacturX_Comfort);

    if (pdfHaveOpenDoc(pdf) != 0) {
        const char* outFile = "out.pdf";
        if (pdfOpenOutputFileA(pdf, outFile) == 0) { pdfDeletePDF(pdf); return 0; }
        if (pdfCloseFile(pdf) != 0)
            printf("PDF file \"%s\" successfully created!\n", outFile);
    }

    pdfDeletePDF(pdf);
    return 0;
}
